pcsx2/GS/GSState.cpp

```diff
@@ -3942,288 +4047,325 @@ static bool UsesRegionRepeat(int fix, int msk, int min, int max, int* min_out, i
+			switch (WM)
+			{
+				case CLAMP_REGION_CLAMP:
+					*min_out = MIN;
+					*max_out = MAX;
+					break;
+				case CLAMP_REGION_REPEAT:
+					*min_out = MIN;
+					*max_out = MIN | MAX;
+					break;
+				default:
+					pxFail("Invalid clamp/wrap mode"); // Impossible
+			}
+		};
 
-	switch (wms)
-	{
-		case CLAMP_REPEAT:
-			break;
-		case CLAMP_CLAMP:
-			break;
-		case CLAMP_REGION_CLAMP:
-			vr.x = minu;
-			vr.z = maxu + 1;
-			break;
-		case CLAMP_REGION_REPEAT:
-			vr.x = maxu;
-			vr.z = (maxu | minu) + 1;
-			break;
-		default:
-			ASSUME(0);
-	}
+		if (WMS == CLAMP_REGION_CLAMP || WMS == CLAMP_REGION_REPEAT)
+			GetRegionBound(WMS, MINU, MAXU, &full_rect_region.left, &full_rect_region.right);
+		if (WMT == CLAMP_REGION_CLAMP || WMT == CLAMP_REGION_REPEAT)
+			GetRegionBound(WMT, MINV, MAXV, &full_rect_region.top, &full_rect_region.bottom);
 
-	switch (wmt)
-	{
-		case CLAMP_REPEAT:
-			break;
-		case CLAMP_CLAMP:
-			break;
-		case CLAMP_REGION_CLAMP:
-			vr.y = minv;
-			vr.w = maxv + 1;
-			break;
-		case CLAMP_REGION_REPEAT:
-			vr.y = maxv;
-			vr.w = (maxv | minv) + 1;
-			break;
-		default:
-			ASSUME(0);
+		full_rect = full_rect.runion(full_rect_region);
 	}
 
-	// Software renderer fixes TEX0 so that TW/TH contain MAXU/MAXV.
-	// Hardware renderer doesn't, and handles it in the texture cache, so don't clamp here.
-	if (clamp_to_tsize)
-		vr = vr.rintersect(tr);
-	else
-		tr = tr.runion(vr);
-
+	GSVector4i vertex_rect; // Rectangle computed from UV bounds of vertices
 	u8 uses_border = 0;
 
+	// FIXME: Replace this with a flag that says "huge or nans" detected in texture coords.
 	if (m_vt.m_max.t.x >= FLT_MAX || m_vt.m_min.t.x <= -FLT_MAX ||
 		m_vt.m_max.t.y >= FLT_MAX || m_vt.m_min.t.y <= -FLT_MAX)
 	{
 		// If any of the min/max values are +-FLT_MAX we can't rely on them
 		// so just assume full texture.
+		vertex_rect = full_rect;
 		uses_border = 0xF;
 	}
 	else
 	{
-		// Optimisation aims to reduce the amount of texture loaded to only the bit which will be read
-		GSVector4 st = m_vt.m_min.t.xyxy(m_vt.m_max.t);
+		// Calculate an approximation of the actual UV range that will be used in the draw
+		// to avoid having to reload the whole texture.
+		GSVector4 vertex_rect_f = m_vt.m_min.t.xyxy(m_vt.m_max.t);
+		
 		if (linear)
 		{
-			st += GSVector4(-0.5f, 0.5f).xxyy();
-			
-			// If it's the start of the texture and our little adjustment is all that pushed it over, clamp it to 0.
-			// This stops the border check failing when using repeat but needed less than the full texture
-			// since this was making it take the full texture even though it wasn't needed.
-			if (!clamp_to_tsize)
-			{
-				const u32 mask = (m_vt.m_min.t.floor() == GSVector4::zero()).mask();
-				if (mask & 1) // X == 0
-					st.x = st.max(GSVector4::zero()).x;
-				if (mask & 2) // Y == 0
-					st.y = st.max(GSVector4::zero()).y;
-			}
+			vertex_rect_f += GSVector4(-0.5f, -0.5f, 0.5f, 0.5f);
```

Comment:
Replacing a lot of this code is going to break a whole bunch of games when the first coordinate is 0, and it goes lower (when it shouldn't) causing the boundary check to wrap.

I hope your new changes don't suffer from this :)